src/zoneinfo/_tzpath.py

```diff
@@ -108,6 +108,58 @@ def _validate_tzfile_path(path, _base=_TEST_PATH):
 del _TEST_PATH
 
 
+def available_timezones():
+    """Returns a set containing all available time zones.
+
+    .. caution::
+
+        This may attempt to open a large number of files, since the best way to
+        determine if a given file on the time zone search path is to open it
+        and check for the "magic string" at the beginning.
+    """
+    from importlib import resources
+
+    valid_zones = set()
+
+    # Start with loading from the tzdata package if it exists: this has a
+    # pre-assembled list of zones that only requires opening one file.
+    try:
+        with resources.open_text("tzdata", "zones") as f:
+            for zone in f:
+                zone = zone.strip()
+                if zone:
+                    valid_zones.add(zone)
+    except (ImportError, FileNotFoundError):
+        pass
+
+    def valid_key(fpath):
+        try:
+            with open(fpath, "rb") as f:
+                return f.read(4) == b"TZif"
+        except Exception:  # pragma: nocover
+            pass
+
+    for tz_root in TZPATH:
+        if not os.path.exists(tz_root):
+            continue
+
+        for root, _, files in os.walk(tz_root):
+            for file in files:
+                fpath = os.path.join(root, file)
+
+                key = os.path.relpath(fpath, start=tz_root)
+                if os.sep != "/":  # pragma: nocover
+                    key = key.replace(os.sep, "/")
+
+                if not key or key in valid_zones:
+                    continue
+
+                if valid_key(fpath):
+                    valid_zones.add(key)
```

Comment:
Just a very small nitpick, this could be simplified as:
```python
if key and valid_key(fpath):
    valid_zones.add(key)
```

Reply:
This is not quite the same thing, because `valid_key` actually opens and reads the file, but it won't actually change the outcome if we've already found the `key` from another source. If `tzdata` is installed, the common case is that we read a single file, populate `valid_zones` with all the standard keys, then start searching `TZPATH` to populate with any additional zones available to the user, so hopefully very few file accesses are actually happening.

Reply:
Ah right, I missed that, sorry.